### PR TITLE
Feature/check sum generator

### PR DESCRIPTION
**Pull request by Joaolfelicio**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Internationalization and localization
- [ ] Other (please describe):

## What is the current behavior?

As a user, I'd like to be able to calculate the checksum of a file and compare it with a hash value so I'm able to know if the file was modified.

Issue Number: #141

## What is the new behavior?

As a user, I'd like to be able to calculate the checksum of a file using multiple hashing algorithms like MD5, SHA1, SHA256, SHA384 or SHA512.

It will display a loading spinner while a file is being hashed unless the file is small, so we don't see the load flicking.

## Other information
main-page

check-sum

check-sum-hashing-algos

check-sum-loading

check-sum-output

## Quality check

Before creating this PR, have you:

- [x] Followed the code style guideline as described in [CONTRIBUTING.md](https://github.com/veler/DevToys/blob/main/CONTRIBUTING.md)
- [x] Verified that the change work in Release build configuration
- [x] Checked all unit tests pass

## What's missing

1. Not sure which font icon/svg to use, open to suggestions!
2. For the moment, the control output comparer is not doing anything, not sure if we want to keep it or what would be the best UX for this control.
3. I did realize that the compact overlay button on the top sometimes throws, could be local as no change related to it was performed.

I will need some help with the above points, thanks.

### CODE REVIEW

**Comment by Joaolfelicio** [comment]
Regarding point number 3, seems like it's only in debug mode!

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+                using HashAlgorithm? hashAlgo = CreateHashAlgorithm(inputHashingAlgorithm);
+                Stream? fileStream = await inputFile.OpenStreamForReadAsync();
+
+                ShouldDisplayProgress = HashingHelper.ComputeHashIterations(fileStream) > _displayAboveIterations;
+
+                byte[]? fileHash = await HashingHelper.ComputeHashAsync(
+                        hashAlgo,
+                        fileStream,
+                        new Progress<HashingProgress>(UpdateProgress),
+                        _cancellationTokenSource.Token);
```
Comment (btiteux) [inline src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs:373]: I see that you use a Stream but I don't see the dispose of this stream?

File: src/dev/impl/DevToys/Models/HashingAlgorithmDisplayPair.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        private HashingAlgorithmDisplayPair(string displayName, HashingAlgorithm value)
+        {
+            DisplayName = displayName;
+            Value = value;
+        }
+
+        public bool Equals(HashingAlgorithmDisplayPair other) => other.Value == Value;
+    }
+
+    public enum HashingAlgorithm
```
[the remark this answers is not in the document]
Comment (Joaolfelicio) [reply src/dev/impl/DevToys/Models/HashingAlgorithmDisplayPair.cs]: Done

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs (hunk 1 above)
Comment (Joaolfelicio) [reply src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs:373]: Well spotted, missed that one 😁

File: src/dev/impl/DevToys/Helpers/HashingHelper.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+using System.Threading.Tasks;
+using DevToys.Models;
+using DevToys.Shared.Core;
+
+namespace DevToys.Helpers
+{
+    internal static class HashingHelper
+    {
+        internal static async Task<byte[]> ComputeHashAsync(
+        HashAlgorithm hashAlgorithm,
```
Comment (veler) [inline src/dev/impl/DevToys/Helpers/HashingHelper.cs:14]: Please indent the parameters by adding 4 spaces.

File: src/dev/impl/DevToys/Helpers/HashingHelper.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        Stream stream,
+        IProgress<HashingProgress> progress,
+        CancellationToken cancellationToken = default,
+        int bufferSize = 1024 * 1024)
+        {
+            Assumes.NotNull(stream, nameof(stream));
```
Comment (veler) [inline src/dev/impl/DevToys/Helpers/HashingHelper.cs:20]: Can you please use `Arguments.NotNull` instead? :)

File: src/dev/impl/DevToys/Helpers/HashingHelper.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace DevToys.Helpers
+{
+    internal static class HashingHelper
+    {
+        internal static async Task<byte[]> ComputeHashAsync(
+        HashAlgorithm hashAlgorithm,
+        Stream stream,
+        IProgress<HashingProgress> progress,
+        CancellationToken cancellationToken = default,
```
Comment (veler) [inline src/dev/impl/DevToys/Helpers/HashingHelper.cs:17]: Ideally, a `CancellationToken` should never be an optional argument. Making it optional encourage future maintainer at bad practice by omitting it.

File: src/dev/impl/DevToys/Helpers/HashingHelper.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+                progress.Report(new HashingProgress(stream.Length, totalBytesRead));
+                cancellationToken.ThrowIfCancellationRequested();
+            }
+            return hashAlgorithm.Hash ?? Array.Empty<byte>();
+        }
+
+        internal static int ComputeHashIterations(Stream stream, int bufferSize = 1024 * 1024)
+        {
+            Assumes.NotNull(stream, nameof(stream));
```
Comment (veler) [inline src/dev/impl/DevToys/Helpers/HashingHelper.cs:55]: Sam here

File: src/dev/impl/DevToys/Helpers/HashingHelper.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
+﻿using System;
```
Comment (veler) [inline src/dev/impl/DevToys/Helpers/HashingHelper.cs:1]: Please add `#nullable enable` at the top of the file.

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+            {
+                Output = Output!.ToLowerInvariant();
+            }
+        }
+
+        private void CheckSum()
+        {
+            lock (_lockObject)
+            {
+                ResetProgress();
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs:325]: I played a little bit with the tool and found an issue in the following scenario:
1. Select a large file.
2. The hash calculation is in progress, the progress ring is increasing. While this time, select another file.

When doing this, the previous calculation process isn't canceled. You might be able to fix it by calling invoking the `CancelCommand` instread of `ResetProgress`.

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+            if (_isCalculationInProgress || InputFile is null)
+            {
+                return;
+            }
+
+            _isCalculationInProgress = true;
+
+            string? hashOutput = await CalculateFileHash(InputHashingAlgorithm.Value, InputFile);
+
+            ThreadHelper.RunOnUIThreadAsync(() =>
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs:341]: Await this method.

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+
+        private async Task CheckSumAsync()
+        {
+            if (_isCalculationInProgress || InputFile is null)
+            {
+                return;
+            }
+
+            _isCalculationInProgress = true;
+
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs:338]: Please call `await TaskScheduler.Default;` here. This allows to force the method to go out of the UI thread, which guarantees us to not hanging the UI until we call RunOnUIThreadAsync

File: src/dev/impl/DevToys/ViewModels/Tools/Graphic/ColorBlindnessSimulator/ColorBlindnessSimulatorToolViewModel.cs [hunk 11]
```diff
@@ -52,10 +52,10 @@ public sealed class ColorBlindnessSimulatorToolViewModel : ObservableRecipient,
 
         internal ColorBlindnessSimulatorStrings Strings => LanguageManager.Instance.ColorBlindnessSimulator;
 
-        internal bool IsSelectFilesAreaHighlithed
+        internal bool IsSelectFilesAreaHighlighted
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Graphic/ColorBlindnessSimulator/ColorBlindnessSimulatorToolViewModel.cs:55]: Good catch! Thank you!

File: src/dev/impl/DevToys/Views/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolPage.xaml [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+            IsReadOnly="True"
+            Text="{x:Bind ViewModel.Output, Mode=OneWay}"/>
+
+        <controls:CustomTextBox
+            Header="{x:Bind ViewModel.Strings.OutputComparer}"
+            Height="50"
+            AcceptsReturn="True"
+            Text="{x:Bind ViewModel.OutputComparer, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}" />
+
+        <controlsToolKit:Loading x:Name="LoadingControl" IsLoading="{x:Bind ViewModel.ShouldDisplayProgress}"  >
```
Comment (veler) [inline src/dev/impl/DevToys/Views/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolPage.xaml:162]: it seems like this isn't used. Should we remove it?

File: src/dev/shared/DevToys.Shared/Core/Assumes.cs [hunk 13]
```diff
@@ -35,5 +35,13 @@ public static void IsTrue(bool value, string argumentName)
                 throw new Exception($"'{argumentName}' is expected to be True");
             }
         }
+
+        public static void NotZeroOrBelow(long value, string argumentName)
```
Comment (veler) [inline src/dev/shared/DevToys.Shared/Core/Assumes.cs:39]: If I'm not wrong, you wanted to use this method to check an argument. If so, can you please move this method to `Arguments` class instead?

**Review by veler — CHANGES_REQUESTED** [review]
Hello,
Thank you very much for this contribution! Overall, it looks very good. Just a few details here and there. :-)

File: src/dev/impl/DevToys/Helpers/HashingHelper.cs (hunk 3 above)
Comment (Joaolfelicio) [reply src/dev/impl/DevToys/Helpers/HashingHelper.cs]: Done

File: src/dev/impl/DevToys/Helpers/HashingHelper.cs (hunk 5 above)
Comment (Joaolfelicio) [reply src/dev/impl/DevToys/Helpers/HashingHelper.cs]: Good point, removed the default value.

File: src/dev/impl/DevToys/Helpers/HashingHelper.cs (hunk 6 above)
Comment (Joaolfelicio) [reply src/dev/impl/DevToys/Helpers/HashingHelper.cs]: Done

File: src/dev/impl/DevToys/Helpers/HashingHelper.cs (hunk 7 above)
Comment (Joaolfelicio) [reply src/dev/impl/DevToys/Helpers/HashingHelper.cs]: Missed this one!

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs (hunk 8 above)
Comment (Joaolfelicio) [reply src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs]: Good catch. Initially was invoking the `CancelCommand` on each calculation, but it was messing up the logic to revert back the dropdown to the previous hashing algorithm if you cancel the calculation.
I'm now passing an argument that specifies if we should revert to the previous hash algorithm or not.
It should be fixed now.

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs (hunk 9 above)
Comment (Joaolfelicio) [reply src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs]: Done.

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs (hunk 10 above)
Comment (Joaolfelicio) [reply src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs:338]: Good point, added it!

File: src/dev/impl/DevToys/Views/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolPage.xaml (hunk 12 above)
Comment (Joaolfelicio) [reply src/dev/impl/DevToys/Views/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolPage.xaml]: Whoopsie! 😅
Removed the control and the import statement.

File: src/dev/shared/DevToys.Shared/Core/Assumes.cs (hunk 13 above)
Comment (Joaolfelicio) [reply src/dev/shared/DevToys.Shared/Core/Assumes.cs]: Done.

**Review by veler — APPROVED** [review]
Awesome! Thank you for the update! It looks good to me. @btiteux , can you please take a look and merge if it looks good to you too?

**Comment by Joaolfelicio** [comment]
Nice! Could you please just take a look at points number 1 and 2?

**Comment by veler** [comment]
> * Not sure which font icon/svg to use, open to suggestions!
> * For the moment, the control output comparer is not doing anything, not sure if we want to keep it or what would be the best UX for this control.

Oh sorry I missed that.

> Not sure which font icon/svg to use, open to suggestions!

It looks good to me. Worst case we can change that later.

> For the moment, the control output comparer is not doing anything, not sure if we want to keep it or what would be the best UX for this control.

That's fine too :)

File: src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs [hunk 14]
```diff
@@ -0,0 +1,432 @@
...
+            await ThreadHelper.RunOnUIThreadAsync(() =>
+            {
+                if (hashOutput != null)
+                {
+                    Output = hashOutput;
+                    _outputHashingAlgorithm = InputHashingAlgorithm;
+                }
+
+                _isCalculationInProgress = false;
+                ShouldDisplayProgress = false;
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Generators/CheckSumGenerator/CheckSumGeneratorToolViewModel.cs:356]: I'm facing a race condition on my machine and I think that this line along with 343 and maybe a few others are the cause.

The scenario I have is:
1. Select a file. Computation start. Progress bar is displayed.
2. Select another file while the computation of the 1st one isn't done yet.
3. In this scenario, the progress bar simply disappear.

I think the race here is that line 343 (with the new file) runs before 356 (with the old file) is executed.

A proper fix is probably to cache the task on line 326 and await it (when not null) right after line 339. Thoughts?

**Pull request merged**